libzmix/src/verifiable_encryption/cs_verifiable_encryption.rs

```diff
@@ -0,0 +1,781 @@
+        let u_sqr =
+            paillier_group.exponentiate(&ciphertext.u, &hs_x3_x2_times_2, Some(&mut ctx))?;
+        let v_sqr = paillier_group.sqr(&ciphertext.v, Some(&mut ctx))?;
+        if v_sqr != u_sqr {
+            return Err(UrsaCryptoError::from_msg(
+                UrsaCryptoErrorKind::InvalidStructure,
+                format!("u^2 != v^2, {:?} != {:?}", &u_sqr, &v_sqr),
+            ));
+        }
+        let mut messages = Vec::<BigNumber>::with_capacity(ciphertext.e.len());
+        for i in 0..ciphertext.e.len() {
+            let u_x1 =
+                paillier_group.exponentiate(&ciphertext.u, &self.pri_key.x1[i], Some(&mut ctx))?;
+            // 1/u^{x_1}
+            let u_x1_inv = u_x1.inverse(&paillier_group.modulus, Some(&mut ctx))?;
+            // (e/u^{x_1})
+            let e_u_x1_inv =
+                &ciphertext.e[i].mod_mul(&u_x1_inv, &paillier_group.modulus, Some(&mut ctx))?;
+
+            // m_hat = (e/u^{x_1})^2*t
+            let m_hat = paillier_group.exponentiate(
+                &e_u_x1_inv,
+                &self.pub_key.two_inv_times_2,
+                Some(&mut ctx),
+            )?;
+            if m_hat.modulus(&self.pub_key.n, Some(&mut ctx))? == *BIGNUMBER_1 {
+                let mut m = m_hat.modulus(&paillier_group.modulus, Some(&mut ctx))?;
+                m.sub_word(1)?;
+                m = m.div(&self.pub_key.n, Some(&mut ctx))?;
+                messages.push(m);
+            } else {
+                return Err(UrsaCryptoError::from_msg(
+                    UrsaCryptoErrorKind::InvalidStructure,
+                    format!("Decryption failed for message {}", i + 1),
+                ));
+            }
+        }
+
+        Ok(messages)
+    }
+
+    /// Encrypt multiple messages.
+    /// Encryption from section 3.2
+    pub fn encrypt(
+        messages: &[BigNumber],
+        label: &[u8],
+        pub_key: &CSEncPubkey,
+    ) -> UrsaCryptoResult<CSCiphertext> {
+        if messages.len() > pub_key.y1.len() {
+            return Err(UrsaCryptoError::from_msg(
+                UrsaCryptoErrorKind::InvalidStructure,
+                format!(
+                    "number of messages {} is more than supported by public key {}",
+                    messages.len(),
+                    pub_key.y1.len()
+                ),
+            ));
+        }
+
+        let paillier_group = &pub_key.paillier_group;
+        let r = paillier_group.rand_for_enc()?;
+
+        Self::encrypt_using_random_value(&r, messages, label, pub_key)
+    }
+
+    /// 1st phase of sigma protocol. Compute ciphertext and commitments (t values).
+    /// Return ciphertext, commitments and random values created during encryption and t value
+    /// "The protocol" from section 5.2. Not using t = g^m*h^s as the idemix protocol does not use it.
+    /// Guess is that since the knowledge of m is proved in the credential attribute proving protocol.
```

Comment:
I did not closely review this section. Would appreciate someone with familiarity with idemix to evaluate the comment about not using t here. Maybe @hartm can more closely review the use of math in this section.

Reply:
@mikelodder7 you resolved this comment. Does that mean you understand and agree with the comment about t?